msal/src/main/java/com/microsoft/identity/client/AccountAdapter.java

```diff
@@ -1,125 +1,296 @@
+            final Map<String, ITenantProfile> tenantProfileMap = new HashMap<>();
+
+            for (final ICacheRecord cacheRecord : entry.getValue()) {
+                final String tenantId = cacheRecord.getAccount().getRealm();
+                final String rawIdToken = getIdToken(cacheRecord);
+                final TenantProfile profile = new TenantProfile(rawIdToken);
+
+                tenantProfileMap.put(tenantId, profile);
+            }
+
+            emptyRoot.setTenantProfiles(tenantProfileMap);
+            result.add(emptyRoot);
+        }
+
+        return result;
+    }
+
+    private static void appendChildren(@NonNull final List<IAccount> rootAccounts,
+                                       @NonNull final List<ICacheRecord> guestCacheRecords) {
+        // Iterate over the roots, adding the children as we go...
+        for (final IAccount account : rootAccounts) {
+            // Iterate over the potential children, adding them if they match
+            final Map<String, ITenantProfile> tenantProfiles = new HashMap<>();
+
+            for (final ICacheRecord guestRecord : guestCacheRecords) {
+                final String guestRecordHomeAccountId = guestRecord.getAccount().getHomeAccountId();
+
+                if (guestRecordHomeAccountId.contains(account.getId())) {
+                    final TenantProfile profile = new TenantProfile(getIdToken(guestRecord));
+                    tenantProfiles.put(guestRecord.getAccount().getRealm(), profile);
+                }
+            }
+
+            // Cast the root account for initialization...
+            final MultiTenantAccount multiTenantAccount = (MultiTenantAccount) account;
+            multiTenantAccount.setTenantProfiles(tenantProfiles);
+        }
+    }
+
+    private static List<IAccount> createRootAccounts(
+            @NonNull final List<ICacheRecord> homeCacheRecords) {
+        final List<IAccount> result = new ArrayList<>();
+
+        for (ICacheRecord homeCacheRecord : homeCacheRecords) {
+            // Each IAccount will be initialized as a MultiTenantAccount whether it really is or not...
+            // This allows us to cast the results however the caller sees fit...
+            final IAccount rootAccount = new MultiTenantAccount(getIdToken(homeCacheRecord));
+
+            // Set the tenant_id
+            ((MultiTenantAccount) rootAccount).setTenantId(
+                    StringUtil.getTenantInfo(
+                            homeCacheRecord
+                                    .getAccount()
+                                    .getHomeAccountId()
+                    ).second
+            );
+
+            // Set the environment...
+            ((MultiTenantAccount) rootAccount).setEnvironment(
+                    homeCacheRecord
+                            .getAccount()
+                            .getEnvironment()
+            );
+
+            result.add(rootAccount);
+        }
+
+        return result;
+    }
+
```

Comment:
nit: @Nullable

Reply:
@rpdome Unless there's something I'm not thinking of these should be `@NonNull` -- I'll annotate it as such

Reply:
Added in d51181ca79ce5837d96fd710d00db4ee846e9b5f